pkg/manifests/external_dns.go

```diff
@@ -143,53 +187,90 @@ func newExternalDNSClusterRole(conf *config.Config, externalDnsConfig *ExternalD
 				Resources: []string{"endpoints", "pods", "services", "configmaps"},
 				Verbs:     []string{"get", "watch", "list"},
 			},
-			{
-				APIGroups: []string{"extensions", "networking.k8s.io"},
-				Resources: []string{"ingresses"},
-				Verbs:     []string{"get", "watch", "list"},
-			},
 			{
 				APIGroups: []string{""},
 				Resources: []string{"nodes"},
 				Verbs:     []string{"get", "watch", "list"},
 			},
 		},
 	}
+
+	switch externalDnsConfig.ResourceType {
+	case ResourceTypeGateway:
+		role.Rules = append(role.Rules,
+			[]rbacv1.PolicyRule{
+				{
+					APIGroups: []string{""},
+					Resources: []string{"namespaces"},
+					Verbs:     []string{"get", "watch", "list"},
+				},
+				{
+					APIGroups: []string{"gateway.networking.k8s.io"},
+					Resources: []string{"gateways", "httproutes", "grpcroutes"},
+					Verbs:     []string{"get", "watch", "list"},
+				},
+			}...,
+		)
+	default:
+		role.Rules = append(role.Rules,
+			rbacv1.PolicyRule{
+				APIGroups: []string{"extensions", "networking.k8s.io"},
+				Resources: []string{"ingresses"},
+				Verbs:     []string{"get", "watch", "list"},
+			})
+	}
+
+	return role
 }
 
 func newExternalDNSClusterRoleBinding(conf *config.Config, externalDnsConfig *ExternalDnsConfig) *rbacv1.ClusterRoleBinding {
-	return &rbacv1.ClusterRoleBinding{
+	var serviceAccount string
+	switch externalDnsConfig.IdentityType {
+	case IdentityTypeWorkloadIdentity:
+		serviceAccount = externalDnsConfig.ServiceAccountName
```

Comment:
this should probably be logic on your new config type. We are mixing responsibilities here and splitting them between this new config type and these functions

Reply:
done, made it an external function.